pallets/funding/src/runtime_api.rs

```diff
@@ -0,0 +1,225 @@
+		let usd_ticket_size = funding_asset_usd_price.saturating_mul_int(asset_amount);
+
+		let mut ct_amount = Zero::zero();
+
+		// Contribution phase
+		if let Some(wap) = project_details.weighted_average_price {
+			ct_amount = wap.reciprocal().unwrap().saturating_mul_int(usd_ticket_size);
+		}
+		// Auction phase, we need to consider multiple buckets
+		else {
+			let mut usd_to_spend = usd_ticket_size;
+			let mut current_bucket = Buckets::<T>::get(project_id).unwrap();
+			while usd_to_spend > Zero::zero() {
+				let bucket_price = current_bucket.current_price;
+				dbg!(&bucket_price);
+				dbg!(&usd_to_spend);
+
+				let ct_to_buy = bucket_price.reciprocal().unwrap().saturating_mul_int(usd_to_spend);
+				dbg!(&ct_to_buy);
+				let ct_to_buy = ct_to_buy.min(current_bucket.amount_left);
+				dbg!(&ct_to_buy);
+
+				ct_amount = ct_amount.saturating_add(ct_to_buy);
+				// if usd spent is 0, we will have an infinite loop
+				let usd_spent = bucket_price.saturating_mul_int(ct_to_buy).max(One::one());
+				dbg!(&usd_spent);
+				usd_to_spend = usd_to_spend.saturating_sub(usd_spent);
+				dbg!(&usd_to_spend);
+				dbg!(&ct_amount);
+
+				current_bucket.update(ct_to_buy)
+			}
+		}
+
+		ct_amount
+	}
+
+	pub fn all_project_participations_by_did(project_id: ProjectId, did: Did) -> Vec<ProjectParticipationIds<T>> {
+		let evaluations = Evaluations::<T>::iter_prefix((project_id,))
+			.filter(|((_account_id, _evaluation_id), evaluation)| evaluation.did == did)
+			.map(|((account_id, evaluation_id), _evaluation)| (account_id, evaluation_id))
+			.collect_vec();
+
+		let bids = Bids::<T>::iter_prefix((project_id,))
+			.filter(|((_account_id, _bid_id), bid)| bid.did == did)
+			.map(|((account_id, bid_id), _bid)| (account_id, bid_id))
+			.collect_vec();
+
+		let contributions = Contributions::<T>::iter_prefix((project_id,))
+			.filter(|((_account_id, _contribution_id), contribution)| contribution.did == did)
+			.map(|((account_id, contribution_id), _contribution)| (account_id, contribution_id))
+			.collect_vec();
+
+		let mut map: BTreeMap<AccountIdOf<T>, (Vec<u32>, Vec<u32>, Vec<u32>)> = BTreeMap::new();
+
+		for (account_id, evaluation_id) in evaluations {
+			map.entry(account_id).or_insert_with(|| (Vec::new(), Vec::new(), Vec::new())).0.push(evaluation_id);
+		}
+
+		for (account_id, bid_id) in bids {
+			map.entry(account_id).or_insert_with(|| (Vec::new(), Vec::new(), Vec::new())).1.push(bid_id);
+		}
+
+		for (account_id, contribution_id) in contributions {
+			map.entry(account_id).or_insert_with(|| (Vec::new(), Vec::new(), Vec::new())).2.push(contribution_id);
+		}
+
+		let output = map
+			.into_iter()
+			.map(|(account, (evaluation_ids, bid_ids, contribution_ids))| ProjectParticipationIds {
+				account,
+				evaluation_ids,
+				bid_ids,
+				contribution_ids,
+			})
+			.collect();
+
+		output
+	}
+
+	pub fn usd_target_percent_reached(project_id: ProjectId) -> FixedU128 {
```

Comment:
To ensure a 1:1 result, shouldn't we reuse the same logic that's used in the `do_end_funding` function? (By abstracting the logic there in a function and calling the same function here and there)